[JSC] Optimize `Number.isSafeInteger` with DFG and FTL for Performance

Gumichocopengin8 opened the pull request:
#### 75f7bb698ee95ce80e8b9c95f816a2d3cdce4e8b
<pre>
[JSC] Optimize `Number.isSafeInteger` with DFG and FTL for Performance
<a href="https://bugs.webkit.org/show_bug.cgi?id=291491">https://bugs.webkit.org/show_bug.cgi?id=291491</a>

Reviewed by Yusuke Suzuki.

Optimize `Number.isSafeInteger` with DFG and FTL for Performance

                        TipOfTree              Patched                Ratio
number-isSafeInteger    0.9238+-0.0566         0.7933+-0.0455         1.1645x faster

* JSTests/microbenchmarks/number-isSafeInteger.js: Added.
* JSTests/stress/number-is-safe-integer.js: Added.
(shouldBe):
(shouldBe.Number.isSafeInteger):
(objectWithNumberValueOf.valueOf):
(objectWithNaNValueOf.valueOf):
(objectRecordConversionCalls.toString):
(objectRecordConversionCalls.valueOf):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGStrengthReductionPhase.cpp:
(JSC::DFG::StrengthReductionPhase::handleNode):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/jit/ThunkGenerators.cpp:
(JSC::numberIsSafeIntegerThunkGenerator):
* Source/JavaScriptCore/jit/ThunkGenerators.h:
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/NumberConstructor.cpp:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::thunkGeneratorForIntrinsic):

Canonical link: <a href="https://commits.webkit.org/296161@main">https://commits.webkit.org/296161@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/102ba1aada05bb90663667a8005b33c6ed2388d1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106603 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26354 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16751 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111811 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/57199 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27024 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34856 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80961 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109607 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/21439 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96199 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61297 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/20893 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14305 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56678 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/99243 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90778 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14335 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114672 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/105221 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33741 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24879 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90024 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34105 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92429 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89733 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34645 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12462 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/29368 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17396 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33666 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/39079 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/129532 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33412 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35283 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/jitless-wasm-to-js-should-follow-to-wasm-call-frame-convention-2.js.default (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36765 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35011 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ab55239a387c4a5aa8457bf5ae94c8b8d298935e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44026

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/23e153aafe17f68a8896b46a54d8f75de4a7c953. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44026

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/110d8220526a9b23bc7b2a2484b37d8d5e83280b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44026

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/bd8207993b2a4821ca6e5e69aa174662a0e80144. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44026

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a6211e6fff6c298930110a085b3322dc186aae3e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44026

sosukesuzuki reviewed (changes requested):
EWS fails

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/deb78b119ca449e337d3faf57cb7da738cd51af1. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44026

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7f2a68e87ca553c913a1347f4ba88bff8f10832d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44026

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a8f30b94895361818e3cfa9d23b315c343781687. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44026

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5c615964a7c57adb475dfe0e3746abda9d3bb20a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44026

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4159df7cd83495716030e60ff1d260c5a1d83418. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44026

Gumichocopengin8 commented:
@sosukesuzuki Do you happen to know why `jsc-armv7-tests` is failing?
The tests pass on macOS, but they fail on Linux. I'm unsure what's causing the issue.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/634c93b716b93b6a272f3c94417f185b90d637fb. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44026

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/53b09d4b2d29fc091a32f17f8e74dc873c2b0c86. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44026

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/005eaa795d73379153a0bb50bc449d5e052e56e0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44026

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/bb70c32dff0d7d7d134d089e2afbe53f51800eb2. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44026

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/596a5e42feb691bc8fa4fce4a0e83e86df7312a0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44026

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ecc40d26a01ea91d04e0d9db1ecca356123fa95b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44026

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6ad3edfb8a28aaefd636d9eba8313e6ea3953f62. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44026

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0953256553710e5045e8413bd36687ee684b9b3d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44026

aoikonomopoulos commented:
Hi @Gumichocopengin8, finally took a look at this. AFAICT the issue on ARMv7 is that `callOperationWithoutExceptionCheck` kills the previously-calculated `isFiniteGPR`. I suspect the easiest way forward is to calculate `isFiniteGPR` after the (potential) call.

Gumichocopengin8 commented:
@aoikonomopoulos Thanks a lot for investigating the issue! I'll try to do either recalculate `isFiniteGPR` or change the order of calculations.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5e249387128289d4567e3f1e895cbc6746fa6294. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44026

Gumichocopengin8 commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -14642,6 +14645,56 @@ IGNORE_CLANG_WARNINGS_END
+            LValue isInRange = m_out.doubleLessThanOrEqual(m_out.doubleAbs(argument), limit);
+
+            LValue constTrue = m_out.constBool(true);
+            LValue isValid = m_out.bitAnd(m_out.doubleEqual(isFinite, constTrue), m_out.doubleEqual(isInteger, constTrue));
```
switched to `equal`

Gumichocopengin8 commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -16874,6 +16874,85 @@ void SpeculativeJIT::compileNumberIsFinite(Node* node)
+        GPRTemporary isInRange(this);
+        FPRTemporary diff(this);
+        FPRTemporary absArgument(this);
+        FPRTemporary limit(this);
```
removed unnecessary FPRs.

Gumichocopengin8 commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -16874,6 +16874,85 @@ void SpeculativeJIT::compileNumberIsFinite(Node* node)
+        GPRTemporary scratch(this);
+        GPRTemporary isFinite(this);
+        GPRTemporary isInteger(this);
+        GPRTemporary isInRange(this);
```
Fixed

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1572c7c2adfa93077b08a7d439a83ec638c7bfb1. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44026

Gumichocopengin8 commented:
@Constellation Thank you for your review. I've updated the PR.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/102ba1aada05bb90663667a8005b33c6ed2388d1. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/44026

Gumichocopengin8 commented on JSTests/stress/number-is-safe-integer.js:
```diff
@@ -0,0 +1,113 @@
+};
+
+shouldBe(Number.isSafeInteger(objectRecordConversionCalls), false);
+shouldBe(objectRecordConversionCalls.callList.length, 0);
```
Fixed

Gumichocopengin8 committed:
[JSC] Optimize `Number.isSafeInteger` with DFG and FTL for Performance
https://bugs.webkit.org/show_bug.cgi?id=291491

Reviewed by Yusuke Suzuki.

Optimize `Number.isSafeInteger` with DFG and FTL for Performance

                        TipOfTree              Patched                Ratio
number-isSafeInteger    0.9238+-0.0566         0.7933+-0.0455         1.1645x faster

* JSTests/microbenchmarks/number-isSafeInteger.js: Added.
* JSTests/stress/number-is-safe-integer.js: Added.
(shouldBe):
(shouldBe.Number.isSafeInteger):
(objectWithNumberValueOf.valueOf):
(objectWithNaNValueOf.valueOf):
(objectRecordConversionCalls.toString):
(objectRecordConversionCalls.valueOf):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGStrengthReductionPhase.cpp:
(JSC::DFG::StrengthReductionPhase::handleNode):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/jit/ThunkGenerators.cpp:
(JSC::numberIsSafeIntegerThunkGenerator):
* Source/JavaScriptCore/jit/ThunkGenerators.h:
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/NumberConstructor.cpp:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::thunkGeneratorForIntrinsic):

Canonical link: https://commits.webkit.org/296161@main
